rust: add a `Task` abstraction.

wedsonaf opened the pull request:
This allows users of the `kernel` crate to get tasks (e.g., the current
task), access their properties (e.g., pid, group_leader, etc.), and hold
on to them if desired. When holding on to tasks, the reference count is
automatically incremented and decremented.

Signed-off-by: Wedson Almeida Filho <wedsonaf@google.com>

ksquirrel commented:
Review of `dab1e55ac67a`:
  - ✔️ Commit [`dab1e55`](https://github.com/Rust-for-Linux/linux/commit/dab1e55ac67a605bda985a12f2a9cc10f052a9b1): Looks fine!

<!-- Review dab1e55ac67a605bda985a12f2a9cc10f052a9b1 -->

ojeda commented on rust/kernel/file.rs:
```diff
@@ -53,7 +53,7 @@ impl Drop for File {
 }
 
-/// A wrapper for [`File`] that doesn't automatically decrement that refcount when dropped.
+/// A wrapper for [`File`] that doesn't automatically decrement the refcount when dropped.
```
I am fine putting the fix in, but as soon as we hit mainline, we will have to be more strict on splitting unrelated fixes (I know you know this -- but stating this for other readers :)

wedsonaf replied:
Agreed. Thanks for accommodating this time around. (I copied/pasted this code and noticed the typo as I adapted it for the new type, so I thought I'd fix the original one too.)

ojeda commented on rust/kernel/task.rs:
````diff
@@ -0,0 +1,188 @@
+/// # }
+/// ```
+///
+/// Getting the pid of the current process, also zero additional cost:
````
Should we use the uppercase, i.e. `PID`, as well as for `OK` etc.? Not sure what is most common (you use `PIDs` below too)

wedsonaf replied:
I was in search of a type that was an acronym in the standard library, the first one I found was `TcpStream` (in golang it would have been TCPStream IIRC), so I followed that.

I have no preference though. Do you think `PID` would be better?

ojeda replied:
Ah, I meant for the comments, not the type names.

wedsonaf replied:
Oh, I see. Yes, I'll change them.

ojeda replied:
`getpid()` manpage and Wikipedia use the uppercase. POSIX uses "process ID". So I guess going with uppercase (for the prose) is better.

ojeda commented on rust/kernel/task.rs:
```diff
@@ -0,0 +1,188 @@
+    pub(crate) ptr: *mut bindings::task_struct,
+}
+
+// SAFETY: Given that the task is referenced, it is ok to send it to another thread.n
```
Spurious `n`.

wedsonaf replied:
Thanks, fixing it.

ojeda commented on rust/kernel/task.rs:
```diff
@@ -0,0 +1,188 @@
+}
+
+impl TaskRef<'_> {
+    /// Constructs a new [`struct task_struct`] wrapper that doesn't change its reference count.
```
Is this link working?

wedsonaf replied:
`rustdoc` doesn't complain about it, but I don't know what it's linking to. Do you think it would be better to remove the square brackets?

ojeda replied:
Perhaps to the bindings? Could you please generate the docs and check the link?

wedsonaf replied:
This function doesn't show up at all. Probably because it's not public. Given that it's not rendered anyway I'll just remove the square brackets.

bjorn3 commented on rust/helpers.c:
```diff
@@ -171,6 +165,24 @@ void rust_helper_rb_link_node(struct rb_node *node, struct rb_node *parent,
 
+struct task_struct *rust_helper_current(void)
+{
+	return current;
```
Is this a global variable or a macro? Of a global variable, it should be accessible as `static` I think.

wedsonaf replied:
It's a macro. It resolves to per-CPU locations (given that each CPU has a different current thread/task), similar in concept to userspace thread-local storage.

nbdd0121 replied:
I think the binding should be named `rust_helper_get_current` and bind to `get_current()`. `current` is basically just `#define current get_current()`.

wedsonaf replied:
> I think the binding should be named `rust_helper_get_current` and bind to `get_current()`. `current` is basically just `#define current get_current()`.

`current` is the idiomatic way to get the current task in C. Do a search for the two and you'll see the difference.

nbdd0121 replied:
I see `current` as a shorthand, and in C that's obvious no reason not to use the shorthand.

However, I think the binding should be named after `get_current` because the binding is a function and it is  used like `let current = rust_helper_get_current();` rather than `let current = rust_helper_current`. The helper could still be implemented with `return current`, I just think the name should change.

If helper generation is made automatic or done like #359, then we should be writing `bindings::get_current()` rather than `bindings::current()`.

wedsonaf replied:
Ok, I'll change the function name to `rust_helper_get_current()`.

bjorn3 commented on rust/kernel/task.rs:
```diff
@@ -0,0 +1,188 @@
+}
+
+// SAFETY: Given that the task is referenced, it is ok to send it to another thread.n
+unsafe impl Send for Task {}
```
`Sync` should be fine too, right?

wedsonaf replied:
Yes, absolutely. I'll add this.

bjorn3 commented on rust/kernel/task.rs:
```diff
@@ -0,0 +1,188 @@
+/// `unsafe` blocks because it would trigger an unbalanced call to `put_task_struct`.
+///
+/// We make this explicitly not [`Send`] so that we can use it to represent the current thread
+/// without having the increment/decrement its reference count.
```
*having to

wedsonaf replied:
Thanks, fixing this.

nbdd0121 commented on rust/kernel/task.rs:
```diff
@@ -0,0 +1,188 @@
+/// The wrapped [`Task`] remains valid for the lifetime of the object.
+pub struct TaskRef<'a> {
+    task: ManuallyDrop<Task>,
+    _not_send: PhantomData<&'a Cell<i32>>,
```
I don't think `Cell` should be used here. Usually `PhantomData<*mut ()>` is used to prevent `Send` and `Sync`. Given that we also need to use `'a` here, how about:
```suggestion
    _not_send: PhantomData<(&'a Task, *mut ())>,
```

wedsonaf replied:
Thanks for the suggestion, I like it.

Though I think `<(&'a (), *mut ())>` would be better, no?

nbdd0121 replied:
I write `&'a Task` purely because `TaskRef` is conceptually a borrowed `Task`. It's phantom anyway so it doesn't really matter ;)

ksquirrel commented:
Review of `6556b37a7fc5`:
  - ✔️ Commit [`6556b37`](https://github.com/Rust-for-Linux/linux/commit/6556b37a7fc57708d820b3779f498f414d130ce8): Looks fine!

<!-- Review 6556b37a7fc57708d820b3779f498f414d130ce8 -->

wedsonaf commented:
Folks, thank you for the prompt reviews! Very much appreciated.

### v1 -> v2

* Renamed `rust_helper_current` to `rust_helper_get_current`.
* Replaced `pid` with `PID` in comments.
* Removed spurious `n` from comment.
* Marked `Task` `Sync` as well.
* Fixed typo in `TaskRef` description (`the` -> `to`).
* Dropped the use of `Cell` to create a non-send phantom variable, using `(&'a (), *mut ())` now.
* Removed brackets around `struct task_struct` in the descriptor of `TaskRef::from_ptr`
